syntax = "proto3";

package connectfour;

service ConnectFour {
    rpc Connect(PlayerInfo) returns (ConnectionStatus);
    rpc MakeMove(Move) returns (GameStatus);
    rpc GetUpdates(PlayerInfo) returns (stream GameUpdate);
}

message PlayerInfo {
    string name = 1;
}

message ConnectionStatus {
    bool connected = 1;
    int32 player_id = 2; // 1 or 2
}

message Move {
    int32 player_id = 1;
    int32 col = 2;
}

message GameStatus {
    bool valid_move = 1;
    string message = 2;
    bool game_over = 3;
    int32 winner = 4; // 0 for no winner, 1 or 2 for player 1 or 2
}

message GameUpdate {
    Board board = 1;
    string message = 2;
    bool game_over = 3;
    int32 winner = 4;
}

message Board {
    repeated Row rows = 1;
}

message Row {
    repeated string cells = 1;
}
